XamarinCommunityToolkit/Behaviors/ImpliedOrderGridBehavior.shared.cs

```diff
@@ -0,0 +1,114 @@
+﻿using System;
+using System.Linq;
+using Xamarin.Forms;
+using Xamarin.Forms.Internals;
+
+namespace Xamarin.CommunityToolkit.Behaviors
+{
+	public class ImpliedOrderGridBehavior : Behavior<Grid>
+	{
+		private bool[][] usedMatrix;
+		private int rowCount;
+		private int columnCount;
+		private Grid attachedGrid;
+
+		public bool ThrowOnLayoutWarning { get; set; }
+
+		protected override void OnAttachedTo(Grid bindable)
+		{
+			base.OnAttachedTo(bindable);
+
+			bindable.ChildAdded += InternalGrid_ChildAdded;
+			attachedGrid = bindable;
+		}
+
+		protected override void OnDetachingFrom(Grid bindable)
+		{
+			base.OnDetachingFrom(bindable);
+
+			bindable.ChildAdded -= InternalGrid_ChildAdded;
+			attachedGrid = null;
+		}
+
+		private void InternalGrid_ChildAdded(object sender, ElementEventArgs e) => 
+			ProcessElement(e.Element);
+
+
+		void LogWarning(string warning)
+		{
+			Log.Warning(nameof(ImpliedOrderGridBehavior), warning);
+			if (ThrowOnLayoutWarning)
+				throw new Exception(warning);
+		}
+
+		bool[][] InitMatrix()
+		{
+			rowCount = attachedGrid.RowDefinitions.Count;
+			if (rowCount == 0)
+				rowCount = 1;
+			columnCount = attachedGrid.ColumnDefinitions.Count;
+			if (columnCount == 0)
+				columnCount = 1;
+			var newMatrix = new bool[rowCount][];
+			for (var r = 0; r < rowCount; r++)
+				newMatrix[r] = new bool[columnCount];
+			return newMatrix;
+		}
+
+		void FindNextCell(out int rowIndex, out int columnIndex)
+		{
+			usedMatrix ??= InitMatrix();
+
+			// Find the first available row
+			var row = usedMatrix.FirstOrDefault(r => r.Any(c => !c));
+
+			// If no row is found, set cell to origin and log
+			if (row == null)
+			{
+				LogWarning("Defined cells exceeded.");
+				rowIndex = rowCount == 0 ? 0 : rowCount - 1;
+				columnIndex = columnCount == 0 ? 0 : columnCount - 1;
+				return;
+			}
+			rowIndex = usedMatrix.IndexOf(row);
+
+			// Find the first available column
+			columnIndex = row.IndexOf(row.First(c => c == false));
```

Comment:
First can raise an exception. Should we use FirstOrDefault and check if value exists?
Also ```boolValue == false``` is preferable to replace by ```!boolValue```